modules/gdscript/gdscript_analyzer.cpp

```diff
@@ -1616,6 +1616,7 @@ void GDScriptAnalyzer::resolve_function_signature(GDScriptParser::FunctionNode *
 		if (!p_is_lambda && get_function_signature(p_function, false, base_type, function_name, parent_return_type, parameters_types, default_par_count, method_flags, &native_base)) {
 			bool valid = p_function->is_static == method_flags.has_flag(METHOD_FLAG_STATIC);
 			valid = valid && parent_return_type == p_function->get_datatype();
+			valid = valid || (!is_type_compatible(p_function->get_datatype(), parent_return_type, true) && is_type_compatible(parent_return_type, p_function->get_datatype()));
```

Comment:
I'd love to see a comment here stating what the now relatively complex boolean is meant to represent! :)

Reply:
I pushed a comment to document the behavior. I also had to change the logic slightly as I introduced a sneaky bug!

The old logic would effectively say:

```cpp
p_function->is_static == method_flags.has_flag(METHOD_FLAG_STATIC) && parent_return_type == p_function->get_datatype()
```
However, by splitting the new condition into a new line, I made it so:
```cpp
p_function->is_static == method_flags.has_flag(METHOD_FLAG_STATIC)
```
can be false and it would still continue checking for covariant. I have to `&&` that static check with the `covariant` check!

Reply:
Nice! Don't forget you can resolve these discussions as well, to clean up the PR discussion! :)

Just one more question, since I'm curious but can't check right now: wouldn't `is_type_compatible(A, B)` imply `!is_type_compatible(B, A)`? That way you could simplify
```valid = valid && (parent_return_type == p_function->get_datatype() || !is_type_compatible(p_function->get_datatype(), parent_return_type, true) && is_type_compatible(parent_return_type, p_function->get_datatype()));```

into:

```valid = valid && (parent_return_type == p_function->get_datatype() || is_type_compatible(parent_return_type, p_function->get_datatype()));```

Reply:
@anvilfolk I am not only doing `!is_type_compatible(B, A)` but `!is_type_compatible(B, A, allow_implicit_conversion = true)` meaning that I don't want the covariant to allow for implicit type conversions but only for non-implicit ones. If for some reason the type could be converted implicitly it should not qualify as a covariant. It might be an unecessary check though, so please correct me if I am mistaken here!

Reply:
Oh, I see what you mean! I don't think I'll allow myself a big opinion here! Hopefully someone with more programming language theory than me will come along and tell us what would feels best here :)